File: android/src/main/java/com/margelo/nitro/nitrocountrypicker/NitroCountryPicker.kt
package com.margelo.nitro.nitrocountrypicker

import android.content.Intent
import androidx.activity.ComponentActivity
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import com.facebook.proguard.annotations.DoNotStrip
import com.facebook.react.bridge.UiThreadUtil
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.NullType
import com.margelo.nitro.core.Promise
import java.util.concurrent.atomic.AtomicInteger


@DoNotStrip
class NitroCountryPicker : HybridNitroCountryPickerSpec() {
  companion object {
    const val NAME = "NitroCountryPicker"

    @Volatile
    private var lastPickedCountry: IPickedCountry? = null

    @Volatile
    private var pendingPickPromise: Promise<Variant_NullType_IPickedCountry>? = null

    private val launcherCounter = AtomicInteger(0)

    private fun setLastPickedCountry(country: IPickedCountry) {
      lastPickedCountry = country
    }

    @Synchronized
    private fun resolvePendingPick(country: IPickedCountry?) {
      val promise = pendingPickPromise ?: return
      pendingPickPromise = null
      val result = if (country == null) {
        Variant_NullType_IPickedCountry.create(NullType.NULL)
      } else {
        Variant_NullType_IPickedCountry.create(country)
      }
      promise.resolve(result)
    }

    @Synchronized
    private fun setPendingPickPromise(promise: Promise<Variant_NullType_IPickedCountry>): Boolean {
      if (pendingPickPromise != null) {
        return false
      }
      pendingPickPromise = promise
      return true
    }

    @Synchronized
    private fun rejectPendingPick(error: Throwable) {
      val promise = pendingPickPromise ?: return
      pendingPickPromise = null
      promise.reject(error)
    }

    private fun nextLauncherKey(): String {
      return "NitroCountryPicker#${launcherCounter.incrementAndGet()}"
    }
  }

  override fun pickCountry(options: PickCountryOptions?): Promise<Variant_NullType_IPickedCountry> {
    val activity = NitroModules.applicationContext?.currentActivity
      ?: return Promise.rejected<Variant_NullType_IPickedCountry>(
        IllegalStateException("No current Activity available to present country picker.")
      )
    val componentActivity = activity as? ComponentActivity
      ?: return Promise.rejected<Variant_NullType_IPickedCountry>(
        IllegalStateException("Current Activity must be a ComponentActivity to register for activity results.")
      )

    val promise = Promise<Variant_NullType_IPickedCountry>()
    val accepted = setPendingPickPromise(promise)
    if (!accepted) {
      return Promise.rejected<Variant_NullType_IPickedCountry>(
        IllegalStateException("Country picker is already open.")
      )
    }

    UiThreadUtil.runOnUiThread {
      var launcher: ActivityResultLauncher<Intent>? = null
      try {
        val intent = Intent(activity, CountryPickerActivity::class.java)
        val headerTitle = options?.headerTitle?.trim()?.takeIf { it.isNotEmpty() }
        if (headerTitle != null) {
          intent.putExtra(CountryPickerActivity.EXTRA_HEADER_TITLE, headerTitle)
        }

        launcher = componentActivity.activityResultRegistry.register(
          nextLauncherKey(),
          ActivityResultContracts.StartActivityForResult()
        ) { result ->
          try {
            val picked = CountryPickerActivity.intentResultToPickedCountry(result.resultCode, result.data)
            if (picked != null) {
              setLastPickedCountry(picked)
            }
            resolvePendingPick(picked)
          } catch (error: Throwable) {
            rejectPendingPick(error)
          } finally {
            launcher?.unregister()
            launcher = null
          }
        }

        launcher?.launch(intent)
      } catch (error: Throwable) {
        try {
          launcher?.unregister()
        } catch (_: Throwable) {
          // Ignore cleanup errors while handling the original failure.
        }
        rejectPendingPick(error)
      }
    }

    return promise
  }

  override fun getLastPickedCountry(): Variant_NullType_IPickedCountry {
    val picked = lastPickedCountry
    return if (picked == null) {
      Variant_NullType_IPickedCountry.create(NullType.NULL)
    } else {
      Variant_NullType_IPickedCountry.create(picked)
    }
  }
}

File: ios/NitroCountryPicker.swift
import CountryPickerAKS
import NitroModules
import React
import UIKit

class NitroCountryPicker: HybridNitroCountryPickerSpec {
  private var lastPickedCountry: IPickedCountry? = nil
  private var pendingPickPromise: Promise<Variant_NullType_IPickedCountry>? = nil

  private func makeError(_ message: String) -> Error {
    return NSError(
      domain: "NitroCountryPicker", code: 1, userInfo: [NSLocalizedDescriptionKey: message])
  }

  @MainActor
  private func resolvePendingPick(country: IPickedCountry?) {
    guard let promise = pendingPickPromise else {
      return
    }
    pendingPickPromise = nil

    let result: Variant_NullType_IPickedCountry = country.map { .second($0) } ?? .first(.null)
    promise.resolve(withResult: result)
  }

  @MainActor
  private func rejectPendingPick(error: Error) {
    guard let promise = pendingPickPromise else {
      return
    }
    pendingPickPromise = nil

    promise.reject(withError: error)
  }

  @MainActor
  private func openPicker(
    options: PickCountryOptions?,
    promise: Promise<Variant_NullType_IPickedCountry>
  ) {
    if pendingPickPromise != nil {
      promise.reject(withError: makeError("Country picker is already open."))
      return
    }
    pendingPickPromise = promise

    guard let presenter = RCTPresentedViewController() else {
      rejectPendingPick(
        error: makeError("No active UIViewController available to present country picker."))
      return
    }

    // CountryPickerAKS does not expose a dedicated header-title config yet.
    _ = options?.headerTitle

    CountryPicker.show(from: presenter) { [weak self] result in
      guard let self else { return }
      switch result {
      case .success(let selected):
        let picked = IPickedCountry(
          name: selected.name,
          dialCode: selected.dial_code,
          code: selected.code
        )
        self.lastPickedCountry = picked
        self.resolvePendingPick(country: picked)
      case .failure(let error):
        switch error {
        case .NotSelected:
          self.resolvePendingPick(country: nil)
        default:
          self.rejectPendingPick(error: error)
        }
      }
    }
  }

  func pickCountry(options: PickCountryOptions?) throws -> Promise<Variant_NullType_IPickedCountry>
  {
    let promise = Promise<Variant_NullType_IPickedCountry>()
    Task { @MainActor [weak self] in
      guard let self else { return }
      self.openPicker(options: options, promise: promise)
    }
    return promise
  }

  func getLastPickedCountry() throws -> Variant_NullType_IPickedCountry {
    if let picked = lastPickedCountry {
      return .second(picked)
    }
    return .first(.null)
  }
}
